neutral_generation/smart_convert.py

```diff
@@ -16,269 +17,303 @@
+    """
+    replace_map = [[token, replacement], [token.capitalize(), replacement.capitalize()],
+                   [token.upper(), replacement.upper()]]
 
-def score(sentence):
+    for j in range(len(replace_map)):
+        pattern = re.compile(r'\b{}\b'.format(replace_map[j][0]))  # \b indicates a word boundary in regex
+        sentence = re.sub(pattern, replace_map[j][1], sentence)
+
+    return sentence
+
+
+def score(sentence: str) -> float:
+    """
+    score the perplexity of a sentence
+    :param sentence: input sentence
+    :return: perplexity normalized by length of sentence (longer sentences won't have inherently have higher perplexity)
+    """
     tokenize_input = tokenizer.tokenize(sentence)
     tensor_input = torch.tensor([tokenizer.convert_tokens_to_ids(tokenize_input)])
     loss = model(tensor_input, lm_labels=tensor_input)
-    return math.exp(loss) / len(tokenize_input)     # normalize perplexity by number of tokens
+    return math.exp(loss) / len(tokenize_input)  # normalize perplexity by number of tokens
+
+
+def simple_replace(token):
+    """
+    mainly deals with straightforward cases of pronoun / gendered word replacement using a lookup
+    also resolves "her" --> "their" / "them"
+    :param token: SpaCy token
+    :return: the token's text replacement (if it exists) as a string.
+    """
+    text = token.text
+
+    # use dependency parser to resolve "her" --> "their" / "them"
+    # if "her" is a possessive pronoun, then its replacement should be "their"
+    # if "her" is an object, then its replacement should be "them"
+    if text.lower() == 'her':
+        is_obj = (token.dep == obj or
+                  token.dep == iobj or
+                  token.dep == pobj or
+                  token.dep == dobj or
+                  token.dep_ == "dative")
+        if token.dep == poss:
+            return capitalization_helper(original=text,
+                                         replacement='their')
+        elif is_obj:
+            return capitalization_helper(original=text,
+                                         replacement='them')
+        else:
+            return None
+
+    # use a lookup for direct mappings
+    # e.g. he --> they, she --> they, policeman --> police officer
+    elif text.lower() in SIMPLE_REPLACE.keys() and token.pos == PRON:
+        replace = SIMPLE_REPLACE[text.lower()]
+        return capitalization_helper(original=text,
+                                     replacement=replace)
+
+    return None
+
+
+def capitalization_helper(original: str, replacement: str) -> str:
+    """
+    helper function to return appropriate capitalization
+    :param original: original word from the sentence
+    :param replacement: replacement for the given word
+    :return: replacement word matching the capitalization of the original word
+    """
+    # check for capitalization
+    if original[0].isupper():
```

Comment:
.istitle()

Reply:
Got it. Thanks for letting me know about this function.